README.md
# ModelSelection
Tools to get an overview of the best model over different sample


## Installation
The package can only be installed from `GitHub` using

```
# check whether devtools are available
# if (!'devtools' %in% installed.packages()) install.packages('devtools')
# install the package
devtools::install_github(repo = 'sknqualitasag/ModelSelection')
```

## GitHub Repository
The repository on GitHub is available at: https://github.com/sknqualitasag/ModelSelection


## Background
One of the first steps in the development of a new genetic analysis routine is to determine the best model for a trait.

## Goal
In the current version, analysis results are read from a txt-file.


## Further Information
Some basic information can be obtained from `carcass_collect_stepAIC_results` article.


NEWS.md
# ModelSelection 0.0.3

* Removed error in read_sample() for csv-file

# ModelSelection 0.0.2

* Adapted Function call in article carcass_collect_stepAIC_results

# ModelSelection 0.0.1

* Added a `NEWS.md` file to track changes to the package.


R/ConvergingResultFiles.R
### #
### # Converging result-files of step backward aic in one result-file
### # 2020-04-27 (skn)
### #-------------------------------------------------------------------


#' Converging result-files of step backward aic in one result-file per sample
#'
#' @param psDirectory
#' @param psOutputFile
#' @export convergingResults_perSample_backAIC
convergingResults_perSample_backAIC <- function(psDirectory,
                                                psOutputFile){

  ### # check parameters
  if(is.null(psDirectory)) {
    stop("--missingDirectory has not been specified")
  }

  ### # Read and combine multiple data files
  convergedFiles <- readbulk::read_bulk(directory = psDirectory,
                                        extension = ".csv",
                                        fun = readr::read_csv)

  ### # Write output file
  readr::write_csv(x = convergedFiles, path = paste(psDirectory,psOutputFile,sep = ""))

}



#' Converging result-files of step backward aic in one overall result-file
#'
#' @param psDirectory
#' @param psSubdirectories
#' @param psOutputFile
#' @export convergingResults_backAIC
convergingResults_backAIC <- function(psDirectory,
                                      psSubdirectories,
                                      psOutputFile){
  ### # check parameters
  if(is.null(psDirectory)) {
    stop("--missingDirectory has not been specified")
  }

  ### # Read and combine multiple data files
  convergedFilePerTrait <- readbulk::read_bulk(directory = psDirectory,
                                               subdirectories = psSubdirectories,
                                               extension = "Overview_ResultPerSample_backAIC.csv",
                                               fun = readr::read_csv)

  ### # Write output file
  readr::write_csv(x = convergedFilePerTrait, path = paste(psDirectory,psOutputFile,sep = ""))

}


R/collect_olsrr_results.R
#' ---
#' title: "Collection of olsrr Result Across Samples"
#' date:  "2020-04-27"
#' ---
#'
#'
#' @title Collect Results from olsrr Modelselection Across Sample Datasets
#'
#' @description
#' For very large datasets model selection is done on smaller subsamples of the
#' complete dataset. Given the subsamples, model selection is run and a set
#' of results is collected. This function is not suitable because the package olsrr
#' can only be used for regression models (models with covariables only).
#' Therefore it does not make sense to use fixed effects (factors) in olsrr model selection.
#'
#' @details
#' Currently model selection is done using \code{olsrr::ols_step_backward_aic()}
#'
#' @return tbl_result tibble with results of the model selection for each sample.
#'
#' @examples
#' \dontrun{
#' n_nr_sample <- 10
#' n_sample_size <- 30
#' # samples from mtcars and write samples to files in tmpdir
#' tmpdir <- tempdir()
#' cat(" * Sample directory: ", tmpdir, "\n")
#' vec_sample_files <- sapply(1:n_nr_sample,
#'                            function(x) paste0("mtcars_sample_", x, ".csv", collapse = ""),
#'                            USE.NAMES = FALSE)
#' vec_sample_path <- file.path(tmpdir, vec_sample_files)
#' set.seed(321)
#' for (i in 1:n_nr_sample){
#'   cat(" * Generate sample ", i, " ...\n")
#'   cur_sample <- dplyr::sample_n(mtcars, n_sample_size, replace = TRUE)
#'   readr::write_csv(cur_sample, path = vec_sample_path[i])
#' }
#' collect_ols_results(pvec_sample = vec_sample_path, ps_response = "mpg")
#' }
#'
#' @param pvec_sample vector of paths to data samples
#' @param ps_response name of the response variable
#' @param pvec_ignoring_columns vector of columns to ignore
#' @return tbl_result tibble with model selection results
#'
#' @importFrom magrittr `%>%`
#'
#' @export collect_ols_results
collect_ols_results <- function(pvec_sample, ps_response, pvec_ignoring_columns = NULL){

  # read the first sample
  tbl_first_sample <- read_sample(ps_sample_path = pvec_sample[1])
  # ignore some columns
  if (!is.null(pvec_ignoring_columns)){
    tbl_first_sample <- tbl_first_sample %>% dplyr::select(-c(pvec_ignoring_columns))
  }
  # check that ps_response is in colnames(tbl_first_sample)
  if (!ps_response %in% colnames(tbl_first_sample))
    stop(" * ERROR: Cannot find response variable: ", ps_response, " in columnnames of first sample.")
  # check that no column is of type factor
  if (any(sapply(colnames(tbl_first_sample), function(x) is.factor(tbl_first_sample[[x]]), USE.NAMES = FALSE)))
    stop(" * ERROR in collect_ols_results: Cannot handle fixed effects - use function collect_stepAIC_results")
  # determine the predictors from column of first sample
  vec_pred <- setdiff(names(tbl_first_sample), ps_response)
  # the vector of column names of the result tibble
  vec_col <- c("SampleID", "(Intercept)", vec_pred, "AIC")

  # determine the number of samples
  n_nr_sample <- length(pvec_sample)
  # initialise the result tibble
  tbl_result <- tibble::as_tibble(matrix(0, nrow = n_nr_sample, ncol = length(vec_col)))
  colnames(tbl_result) <- vec_col

  # specify the formula for the full model
  formula.full <- as.formula(paste( ps_response, paste0(vec_pred, collapse = " + "), sep = " ~ "))

  # loop over the samples and do the model selection
  for (sidx in seq_along(pvec_sample)){
    tbl_cur_sample <- read_sample(ps_sample_path = pvec_sample[sidx])
    # check whether response variable is in colnames of tbl_cur_sample
    if (!ps_response %in% colnames(tbl_cur_sample))
      stop(" * ERROR: Cannot find response variable: ", ps_response, " in columnnames of current sample.")
    # check that no column is of type factor
    if (any(sapply(colnames(tbl_cur_sample), function(x) is.factor(tbl_cur_sample[[x]]), USE.NAMES = FALSE)))
      stop(" * ERROR in collect_ols_results: Cannot handle fixed effects - use function collect_stepAIC_results")

    # fit full model to current sample data
    lm.full.cur <- lm(formula = formula.full, data = tbl_cur_sample)
    ols_result <- olsrr::ols_step_backward_aic(model = lm.full.cur)
    tbl_result[sidx, names(coefficients(ols_result$model))] <- 1
    tbl_result[sidx, "AIC"] <- ols_result$aics[length(ols_result$aics)]
    tbl_result[sidx, "SampleID"] <- sidx
  }

  # return the resulting tibble
  return(tbl_result)
}


R/Run_ModelSelectionCarcass.R
### #
### # Running Model Selection for carcass calves and adults
### # 2020-04-27 (skn)
### # ---------------------------------------------


#' @title Running model selection for carcass
#'
#' @param psInputFile
#' @param psOutputPath
#' @param psSampleID
#' @export running_ms_carcass
running_ms_carcass <- function(psInputFile,
                               psOutputPath,
                               psSampleID){

  suppressPackageStartupMessages(if(! require("dplyr")) {
    install.packages("dplyr", repos="https://stat.ethz.ch/CRAN/")
    require("dplyr")
  })

  ### # check parameters
  if(is.null(psInputFile)) {
    stop("--missingDataFile has not been specified")
  }

  ### # Read dataset
  tbl_data <- readr::read_delim(file = psInputFile, delim = " ")

  ### # Remove and rename some columns
  tbl_data <- tbl_data %>% select(-c(herdyear,herd,calfcull,controlcull,ageVK,slaughterCategory,branding,ccvn,ccvo,cwvn,cwvo,cfvn,cfvo,cavn,cavo))
  names(tbl_data)[30] <- "het"
  names(tbl_data)[31] <- "rec"

  ### # Convert all fixed effects into factors
  tbl_data$sex <- as.factor(tbl_data$sex)
  tbl_data$abattoir <- as.factor(tbl_data$abattoir)
  tbl_data$classifier <- as.factor(tbl_data$classifier)
  tbl_data$yearsaison <- as.factor(tbl_data$yearsaison)
  tbl_data$breedcomb <- as.factor(tbl_data$breedcomb)
  tbl_data$prodsystem <- as.factor(tbl_data$prodsystem)
  tbl_data$het <- as.factor(tbl_data$het)
  tbl_data$rec <- as.factor(tbl_data$rec)

  ### # All trait without covariable age and ageQuadrat
  Merkmal_oCov <- c("cccn","ccco","ccan","ccas","ccao",
                    "cwcn","cwco","cwan","cwas","cwao",
                    "cfcn","cfco","cfan","cfas","cfao",
                    "cacn","caco","caan","caas","caao")

  for(mrkml in Merkmal_oCov){
    cat("mrkml: ", mrkml,"\n")
    ModelSelection::generate_output_backAIC(psDataFile = tbl_data,
                                            psTrait = mrkml,
                                            psOutputDir = psOutputPath,
                                            psSampleNUM = psSampleID,
                                            psTraitComment = "oCov")
  }

  ### # All trait with covariable age and ageQuadrat
  Merkmal_mCov <- c("cccn","ccco","ccan","ccas","ccao",
                    "cwcn","cwco","cwan","cwas","cwao",
                    "cfcn","cfco","cfan","cfas","cfao")


  for(mrkml in Merkmal_mCov){
    ModelSelection::generate_output_backAIC(psDataFile = tbl_data,
                                            psTrait = mrkml,
                                            psOutputDir = psOutputPath,
                                            psSampleNUM = psSampleID,
                                            psTraitComment = "mCov")
  }

  ### # Output with all result of a sample
  ModelSelection::convergingResults_perSample_backAIC(psDirectory = psOutputPath,
                                                      psOutputFile = "Overview_ResultPerSample_backAIC.csv")


}








R/GeneratorOutputBackwardAIC.R
### #
### # Generate an output with result of step backward aic for each model
### # 2020-04-24 (skn)
### #-------------------------------------------------------------------


#' @title Generate an output for result of step backward for carcass
#'
#' @param psDataFile
#' @param psTrait
#' @param psOutputDir
#' @param psSampleNUM
#' @param psTraitComment
#' @export generate_output_backAIC
generate_output_backAIC <- function(psDataFile,
                                    psTrait,
                                    psOutputDir,
                                    psSampleNUM,
                                    psTraitComment){

  suppressPackageStartupMessages(if(! require("dplyr")) {
    install.packages("dplyr", repos="https://stat.ethz.ch/CRAN/")
    require("dplyr")
  })

  ### # check parameters
  if(is.null(psDataFile)) {
    stop("--missingDataFile has not been specified")
  }

  # Build a model
  # With age as covariable
  if(psTraitComment == "mCov"){
    lm_model <- lm(as.formula(paste(psTrait, "~ sex + abattoir + classifier + yearsaison + breedcomb + het + rec + age + ageQuadrat")), data = psDataFile)
  }else{
    #Without age as covariable
    lm_model <- lm(as.formula(paste(psTrait, "~ sex + abattoir + classifier + yearsaison + breedcomb + het + rec")), data = psDataFile)
  }

  # Check the model with the function 'ols_step_backward_aic' of the Rpackage olsrr
  k <- olsrr::ols_step_backward_aic(lm_model)
  model <- paste(format(lm_model$terms)[1],format(lm_model$terms)[2],sep = "")
  optimumaic <- k$aics[length(k$predictors)+1]

  # Build a tibble as output
  # With age as covariable
  if(psTraitComment == "mCov"){
    tibble_output <- tibble::tibble(psSampleNUM, psTrait, psTraitComment, model, AIC = optimumaic, sex = 1, abattoir = 1, classifier = 1,
                                    age = 1, ageQuadrat = 1, yearsaison = 1, breedcomb = 1, het = 1, rec = 1)
  }else{
    #Without age as covariable
    tibble_output <- tibble::tibble(psSampleNUM, psTrait, psTraitComment, model, AIC = optimumaic, sex = 1, abattoir = 1, classifier = 1,
                                    age = NA, ageQuadrat = NA, yearsaison = 1, breedcomb = 1, het = 1, rec = 1)
  }

  # Update tibble with variables which are removed
  if(!is.null(k$predictors)){
    i <- 1
    for(i in 1:length(k$predictors)){
      variable2remove <- k$predictors[i]
      tibble_output[,variable2remove] <- 0
      i <- i+1
    }
  }

  # Write tibble-Output in a file
  readr::write_csv(x = tibble_output, path = paste(psOutputDir,paste(psSampleNUM,psTrait,psTraitComment,sep = "_"),".csv",sep = ""))

}


inst2/scratch/CheckHeterosisCode.R
### #
### # Check Heterosis Codierung vor Optimierung
### # 2020-03-06 (skn)
### # -------------------------------------------------

### # Libraries
library(dplyr)
library(ggplot2)

### # Read Output codeEffHetRec
s_het <- "../../work/Sample_ForModellSelektion_Produktionssystem_Het/fleisch.out"
tbl_het <- readr::read_delim(file = s_het, delim = " ")
tbl_het <- tbl_het %>% select(c("ind","cccn","ccco","ccan","ccas","ccao","cwcn","cwco","cwan","cwas","cwao","cfcn","cfco","cfan","cfas","cfao","Effekt_HeterosisSumHet","Effekt_RekombinationSumRec"))
rm(s_het)

### # Overview number per productionsystem
tbl_het %>% filter(cccn != 0) %>% count()
tbl_het %>% filter(ccco != 0) %>% count()
tbl_het %>% filter(ccan != 0) %>% count()
tbl_het %>% filter(ccas != 0) %>% count()
tbl_het %>% filter(ccao != 0) %>% count()

### # Read Output calcHetRec
s_calcHet <- "../../work/Sample_ForModellSelektion_Produktionssystem_Het/hetRecRed.csv"
tbl_calcHet <- readr::read_delim(file = s_calcHet, delim = ";")
rm(s_calcHet)

### # Merge both file
tbl_heterosis <- tbl_calcHet %>% inner_join(tbl_het, by = c("TierID_TVD" = "ind"))
rm(tbl_het);rm(tbl_calcHet)

### # Plot Heterosis again Y
ggplot(tbl_heterosis, aes(x=ccao, y=SumHet)) +
  geom_point()

ggplot(tbl_heterosis, aes(x=cfao, y=SumHet)) +
  geom_point()

ggplot(tbl_heterosis, aes(x=cwao, y=SumHet)) +
  geom_point()


R/collect_stepAIC_results.R
#' ---
#' title: "Collection of MASS::stepAIC Results"
#' date:  "2020-04-29"
#' #' ---
#'
#'
#' @title Collect Results From MASS::stepAIC Across Sample Datasets
#'
#' @description
#' Model selection is done with \code{MASS::stepAIC()} for a set of sample
#' datasets. The results are collected into a tibble and returned to the
#' caller.
#'
#'#' @examples
#' \dontrun{
#' n_nr_sample <- 10
#' n_sample_size <- 30
#' # samples from mtcars and write samples to files in tmpdir
#' tmpdir <- tempdir()
#' cat(" * Sample directory: ", tmpdir, "\n")
#' vec_sample_files <- sapply(1:n_nr_sample,
#'                            function(x) paste0("mtcars_sample_", x, ".csv", collapse = ""),
#'                            USE.NAMES = FALSE)
#' vec_sample_path <- file.path(tmpdir, vec_sample_files)
#' set.seed(321)
#' for (i in 1:n_nr_sample){
#'   cat(" * Generate sample ", i, " ...\n")
#'   cur_sample <- dplyr::sample_n(mtcars, n_sample_size, replace = TRUE)
#'   readr::write_csv(cur_sample, path = vec_sample_path[i])
#' }
#' collect_stepAIC_results(pvec_sample = vec_sample_path, ps_response = "mpg")
#' }
#'
#' @param pvec_sample vector of paths to sample files
#' @param ps_response response variable
#' @param pvec_ignoring_columns columns to be ignored in sample data
#' @param pvec_fixeffect_columns columns to be treated as fixed effects
#' @param pb_trace flag to indicate whether stepAIC produced output
#' @return tbl_result tibble with model selection results
#'
#' @importFrom magrittr `%>%`
#'
#' @export collect_stepAIC_results
collect_stepAIC_results <- function(pvec_sample, ps_response,
                                    pvec_ignoring_columns = NULL,
                                    pvec_fixeffect_columns = NULL,
                                    pb_trace = FALSE){
  # read first sample file
  tbl_first_sample <- read_sample(ps_sample_path = pvec_sample[1])
  # ignore some columns
  if (!is.null(pvec_ignoring_columns)){
    tbl_first_sample <- tbl_first_sample %>% dplyr::select(-c(pvec_ignoring_columns))
  }
  # check that ps_response is in colnames(tbl_first_sample)
  if (!ps_response %in% colnames(tbl_first_sample))
    stop(" * ERROR in collect_stepAIC_results: Cannot find response variable: ", ps_response, " in columnnames of first sample.")

  # determine the predictors from column of first sample
  vec_pred <- setdiff(names(tbl_first_sample), ps_response)
  # the vector of column names of the result tibble
  vec_col <- c("SampleID", vec_pred, "AIC")

  # determine the number of samples
  n_nr_sample <- length(pvec_sample)
  # initialise the result tibble
  tbl_result <- tibble::as_tibble(matrix(0, nrow = n_nr_sample, ncol = length(vec_col)))
  colnames(tbl_result) <- vec_col

  # specify the formula for the full model
  formula.full <- as.formula(paste( ps_response, paste0(vec_pred, collapse = " + "), sep = " ~ "))
  # loop over the samples and do the model selection
  for (sidx in seq_along(pvec_sample)){
    # reading the current dataset
    tbl_cur_sample <- read_sample(ps_sample_path =  pvec_sample[sidx])
    # check that ps_response is in colnames(tbl_cur_sample)
    if (!ps_response %in% colnames(tbl_cur_sample))
      stop(" * ERROR in collect_stepAIC_results: Cannot find response variable: ", ps_response, " in columnnames of current sample.")
    # ps_response assure that have a record
    tbl_cur_sample <- tbl_cur_sample[tbl_cur_sample[[ps_response]] != 0,]
    # Distinguish between fix effects (as.factor) and covariables (still numeric)
    if (!is.null(pvec_fixeffect_columns)){
      for(fx in pvec_fixeffect_columns){
        tbl_cur_sample[[fx]] <- as.factor(tbl_cur_sample[[fx]])
        # check whether there are sufficient number of levels in the fixed effect fx
        if (nlevels(tbl_cur_sample[[fx]]) < 2)
          stop(" * ERROR in collect_stepAIC_results: fixed effect ", fx, " in dataset ", pvec_sample[sidx], " has insufficient number of levels: ", nlevels(tbl_cur_sample[[fx]]))
      }
    }
    lm.full.cur <- lm(formula = formula.full, data = tbl_cur_sample)
    if(AIC(lm.full.cur) == -Inf)
      stop(" *ERROR in collect_stepAIC_results: AIC is -Inf")
    step_result <- MASS::stepAIC(lm.full.cur, direction = 'backward', trace = pb_trace)
    tbl_result[sidx, labels(terms(step_result))] <- 1
    tbl_result[sidx, "AIC"] <- step_result$anova$AIC[length(step_result$anova$AIC)]
    tbl_result[sidx, "SampleID"] <- sidx
  }

  # return the resulting tibble
  return(tbl_result)
}


#' --- Reading Sample files ---
#'
#' @title Read a Sample File
#'
#' @description
#' The function distinguishes between files with extension "csv" and "txt" or "out".
#' Sample files with extension "csv" are read with readr::read_csv and sample files
#' with extension "txt" or "out" are read with readr::read_delim. Sample files with
#' extensions other than "csv", "txt" or "out" are not read and produce an error.
#'
#' @param ps_sample_path path to sample file
#' @return tbl_sample tibble read from sample file
read_sample <- function(ps_sample_path){
  #depending of the file-extension, different reading functions for the first sample
  if(assertthat::has_extension(path = ps_sample_path, ext = "csv")){
    tbl_sample <- readr::read_delim(file = ps_sample_path, delim = ";")
  } else if(assertthat::has_extension(path = ps_sample_path, ext = "txt") ||
            assertthat::has_extension(path = ps_sample_path, ext = "out")){
    tbl_sample <- readr::read_delim(file = ps_sample_path, delim = " ", col_types = readr::cols())
  } else{
    stop(" * ERROR in read_sample: Cannot find a defined file-extension of the first file: ", ps_sample_path)
  }
  return(tbl_sample)
}
